File: Data Structures & Algorithms/count-number-of-islands/submission-0.cpp

class Solution {
public:
  void bfs(vector<vector<char>>& grid, vector<vector<bool>>& visited, int r, int c) {
    queue<pair<int, int>> q;
    int r_limit = visited.size();
    int c_limit = visited[0].size();
    q.push({r, c});

    while (q.size() > 0) {
      auto [row, col] = q.front();
      q.pop();
      vector<pair<int, int>> directions = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

      for (auto[dr, dc] : directions) {
        int rl = row + dr;
        int cl = col + dc;
           if (rl >= 0 && rl < r_limit && cl >= 0 && cl < c_limit && !visited[rl][cl] && grid[rl][cl] == '1')
           {
            visited[rl][cl] = 1;
            q.push({rl, cl});
           }
        
           
      }
    }
  }
  int numIslands(vector<vector<char>> &grid) {
    int r = grid.size();
    int c = grid[0].size();
    int islands = 0;
  vector<vector<bool>> visited(r, vector<bool>(c, false));

    for (int i = 0; i < r; i++) {
      for (int j = 0; j < c; j++) {
        if (grid[i][j] == '1' && !(visited[i][j]))
        {
          bfs(grid, visited, i, j);
          islands++;
          visited[i][j] = 1;
        }
      }
    }
    return islands;
  }
};

File: Data Structures & Algorithms/max-area-of-island/submission-0.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        visited = set()
        maxArea = 0
        
        def bfs(i, j):
            tmp = 1
            
            queue = deque()
            queue.append((i, j))

            while(len(queue) > 0):
                r, c = queue.popleft()
                directions = {(1, 0), (-1, 0), (0, 1), (0, -1)}

                for dr, dc in directions: 
                    row = dr + r
                    col = dc + c
                    if 0 <= row < len(grid) and 0 <= col < len(grid[0]) and (row, col) not in visited and grid[row][col] == 1:
                        visited.add((row, col))
                        tmp += 1
                        queue.append((row, col))
            
            return tmp

        

        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if(grid[i][j] == 1 and  (i, j) not in visited):
                    visited.add((i, j))
                    maxArea = max(maxArea, bfs(i, j))
                    
                    

        return maxArea
